controllers/utils.js

```diff
@@ -33,71 +33,161 @@ function hasUserSignedUpForRelay(user) {
   }
 
-  if (req.session.excludeFromExperiment && !req.query.experimentBranch) {
+  // make sure random number falls in the distribution range
+  let runningTotal;
+  let choice;
+
+  if (sorterNum <= totalPercentage) {
+      runningTotal = 0;
+
+      // loop through all variations
+      for (const v in variations) {
+          // check if random number falls within current variation range
+          if (sorterNum <= (variations[v] + runningTotal)) {
+              // if so, we have found our variation
+              choice = v;
+              break;
+          }
+
+          // tally variation percentages for the next loop iteration
+          runningTotal += variations[v];
+      }
+    }
+
+    return choice;
+
+}
+
+function getExperimentBranch(req, sorterNum = false, language = false, variations) {
+
+  const session = req.session.experimentFlags;
+
+  if (session.excludeFromExperiment && !req.query.experimentBranch) {
     log.debug("This session has already been excluded from the experiment");
+    session.excludeFromExperiment = true;
+    session.experimentBranch = false;
+    session.treatmentBranch = false;
     return false;
   }
 
   // If we cannot parse req.headers["accept-language"], we should not
   // enroll users in the experiment.
   if (language && !req.headers || language && !req.headers["accept-language"]){
     log.debug("No headers or accept-language information present.");
+    session.excludeFromExperiment = true;
+    session.experimentBranch = false;
+    session.treatmentBranch = false;
     return false;
   }
 
   // If the user doesn't have the requested variant langauge selected as their primary language,
   // we do not enroll them in the experiment.
   if (language) {
+
+    if (!Array.isArray(language)) {
+      throw new Error("The language param is not an array");
+    }
     const lang = req.headers["accept-language"].split(",");
-    if (language && !lang[0].includes(language)) {
-      log.debug(`Preferred language is not ${language} variant: ${lang[0]}`);
+
+    // Check to make sure one of the experiment langauge(s) is the top-preferred language.
+    const firstLangMatch = (element) => lang[0].includes(element);
+    if (language && !language.some(firstLangMatch)) {
+      log.debug(`Preferred language is not [${language}] variant: ${lang[0]}`);
+      session.excludeFromExperiment = true;
+      session.experimentBranch = false;
+      session.treatmentBranch = false;
       return false;
     }
   }
 
   // If URL param has experimentBranch entry, use that branch;
   if (req.query.experimentBranch) {
-    if (!["va", "vb"].includes(req.query.experimentBranch)) {
+    if (!Object.keys(variations).includes(req.query.experimentBranch)) {
       log.debug("The requested branch is unknown: ", req.query.experimentBranch);
+      session.excludeFromExperiment = true;
+      session.experimentBranch = false;
+      session.treatmentBranch = false;
       return false;
```

Comment:
DRY Check: Anywhere it returns false, we set these three vars. May make sense to put this as a separate function?

Reply:
Yeah, could be this easy?
```js
return unEnrollSession(session);
...
function unEnrollSession(session) {
    session.excludeFromExperiment = true;
    session.experimentBranch = false;
    session.treatmentBranch = false;
    return false;
}
```